store/migration/stream.go

```diff
@@ -0,0 +1,70 @@
+package migration
+
+import (
+	"fmt"
+	"io"
+	"sync/atomic"
+
+	protoio "github.com/cosmos/gogoproto/io"
+	"github.com/cosmos/gogoproto/proto"
+
+	"cosmossdk.io/store/v2/snapshots"
+	snapshotstypes "cosmossdk.io/store/v2/snapshots/types"
+)
+
+// MigrationStream is a stream for migrating the whole state as a snapshot.
+// It's used to sync the whole state from the store/v1 to store/v2.
+// The main idea is to use the same snapshotter interface without writing to disk.
+type MigrationStream struct {
+	chBuffer chan proto.Message
+	err      atomic.Value // atomic error
+}
+
+var (
+	_ snapshots.WriteCloser = (*MigrationStream)(nil)
+	_ protoio.ReadCloser    = (*MigrationStream)(nil)
+)
+
+// NewMigrationStream returns a new MigrationStream.
+func NewMigrationStream(chBufferSize int) *MigrationStream {
+	return &MigrationStream{
+		chBuffer: make(chan proto.Message, chBufferSize),
+	}
+}
+
+// WriteMsg implements protoio.Write interface.
+func (ms *MigrationStream) WriteMsg(msg proto.Message) error {
+	ms.chBuffer <- msg
+	return nil
+}
+
+// CloseWithError implements snapshots.WriteCloser interface.
+func (ms *MigrationStream) CloseWithError(err error) {
+	ms.err.Store(err)
+	close(ms.chBuffer)
+}
+
+// ReadMsg implements protoio.Read interface.
+func (ms *MigrationStream) ReadMsg(msg proto.Message) error {
+	snapshotsItem, ok := msg.(*snapshotstypes.SnapshotItem)
+	if !ok {
+		return fmt.Errorf("unexpected message type: %T", msg)
+	}
+	item := <-ms.chBuffer
+	if item == nil {
+		return io.EOF
+	}
+	*snapshotsItem = *(item.(*snapshotstypes.SnapshotItem))
+
+	err := ms.err.Load()
```

Comment:
we only have a possible error when `CloseWithError` is called. So why check it here?

Reply:
`CloseWithError` is triggered on the writer side. It is used to let the reader know about the error of the writer side.